src/utilities/modules/temperatureSensorHelpers.js

```diff
@@ -370,87 +378,119 @@ function addHeadAndTrailingLogToBreach({ result, sensor, database }) {
+ * maximum of 24 sensor logs for each full aggregate log.
+ * If within the 8 hour period a breach occurred, there will be less sensorlogs,
+ * to a minimum of 1.
+ * @param {object}       result   object containing the result of aggregation
+ * @param {Realm.Sensor} sensor   The Sensor object for which the logs should be aggregated
+ * @param {Realm}        database App-wide database entry point
+ */
+function doFullAggregation({ result, sensor, database }) {
+  const { sensorLogs } = sensor;
+  if (!sensorLogs || sensorLogs.length === 0) return null;
+  // Want to leave the last 2 hours of logs un-aggregated. Filter them
+  // out.
+  const maxTimeToAggregate =
+    sensorLogs.filtered('aggregate == $0', 'preAggregate').max('timestamp') -
+    TWO_HOURS_MILLISECONDS;
+  const sortedSensorLogs = sensorLogs
+    .filtered('aggregate == $0 && timestamp < $1', 'preAggregate', maxTimeToAggregate)
+    .sorted('timestamp');
+  // Initialized for creating a timestamp of the first sensor log for
+  // a sequential group of logs, which are delimited by a sensorlog
+  // outside of the 8 hour window this group is aggregated for.
+  let logGroupStartTimestamp = null;
+  // 2D array for all groups of to-be-aggregated sensor logs.
+  const sensorLogGroups = [];
+  // A collection of sensor logs, grouped for the currently being operated on,
+  // < 8 hour period.
+  let sensorLogGroup = [];
+  sortedSensorLogs.forEach(sensorLog => {
+    const { timestamp } = sensorLog;
+    // Premature return if this sensorlog is malformed
+    if (!timestamp) return;
+    // Check if this sensorLog is in the current interval.
+    const isInSameInterval = logGroupStartTimestamp - timestamp < FULL_AGGREGATION_INTERVAL;
+    // If so, simply add it to the current group of logs.
+    if (isInSameInterval) sensorLogGroup.push(sensorLog);
+    // Otherwise, hit a delimiter. If there are already logs in the current
+    // sensor log group, this is an ending delimiter, store the group and reset.
+    else if (sensorLogGroup.length !== 0) sensorLogGroups.push([...sensorLogGroup]);
+    // Resetting for a new group of sensor logs;
+    if (!isInSameInterval) {
+      sensorLogGroup = [sensorLog];
+      logGroupStartTimestamp = timestamp;
+    }
   });
-
+  // Create an array of aggregated sensor log objects
+  const aggregatedSensorLogs = sensorLogGroups
+    .map(group => createFullAggregateSensorLogs(group))
+    .flat();
+  // Store each aggregated sensorlog in the database.
   database.write(() => {
-    logsToAdd.forEach(sensorLog => {
-      const newSensorLog = database.update('SensorLog', sensorLog);
-      sensor.sensorLogs.push(newSensorLog);
-      linkToBatches(newSensorLog, database);
-    });
+    aggregatedSensorLogs.forEach(aggregatedLog => database.update('SensorLog', aggregatedLog));
+    // Delete each pre-aggregated sensorlog.
+    sortedSensorLogs.forEach(preAggregatedLog => database.delete(preAggregatedLog));
```

Comment:
was wondering, because sortedSensorLog is realm collection, is it safe to iterate it while deleting the logs ? i.e. would it mutate during iteration causing adverse effects ?